pkg/common/fflag/fflag.go

```diff
@@ -0,0 +1,126 @@
+// The fflag package implements a basic singleton pattern for the purpose of
+// providing SPIRE with a system-wide feature flagging facility. Feature flags
+// can be easily added here, in a single central location, and be consumed
+// throughout the codebase.
+package fflag
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+// Flag represents a feature flag and its configuration name
+type Flag string
+
+// To add a feature flag, declare it in `common.go`, `agent.go`, or
+// `server.go` depending on where it should be considered valid. Then,
+// add it to the relevant package-level map, setting the appropriate
+// default value. Common feature flags are added to the map in `common.go`,
+// while agent or server specific flags are added to the map in the `afflag` or
+// `sfflag` package, located in the agent or server codebase, respectively.
+// Flags should always be opt-in, and default to false, with the only exception
+// being flags that are in the process of being deprecated.
+var (
+	singleton = struct {
+		flags  map[Flag]bool
+		loaded *sync.Once
+		mtx    *sync.RWMutex
+	}{
+		flags:  commonFlagMap,
+		loaded: new(sync.Once),
+		mtx:    new(sync.RWMutex),
+	}
+)
+
+// Load initializes the fflag package and configures its feature flag state
+// based on the configuration input. Feature flags are designed to be
+// Write-Once-Read-Many, and as such, Load can be called only once. Load will
+// return an error if it is called more than once, if the configuration input
+// cannot be parsed, or if an unrecognized flag is set.
+func Load(rc RawConfig, supplementalFlags map[Flag]bool) error {
+	flagConfig, err := parseRawConfig(rc)
+	if err != nil {
+		return fmt.Errorf("could not parse feature flag configuration: %w", err)
+	}
+
+	err = validateFlags(flagConfig, supplementalFlags)
+	if err != nil {
+		return fmt.Errorf("bad feature flag configuration: %w", err)
+	}
+
+	ok := false
+	singleton.loaded.Do(func() {
+		ok = true
+		load(flagConfig, supplementalFlags)
+	})
```

Comment:
> nit: validateFlags and load are a little harder to reason about since they both touch global state (i.e. singleton.flags). I think the code would be easier to follow (and shorter) if they were combined and more functional

I'll have a look at this and see what I can do

> func load(config map[string]bool, flagSets ...map[Flag]bool) (map[Flag]bool, error)

I don't think we have any use cases for loading multiple flag sets beyond just common and agent/server? My preference on this work is to just build what we need right now

> Where the flagsSets are presented in descending priority order.

We can move to error condition so there is no priority

Reply:
That suggested interface was meant to generalize just for convenience iterating over the flag sets inside the function, since the inner loop that processes each flag would be the same. I think having a distinct parameter for both supplemental and common is fine as well.